h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -71,73 +104,39 @@ private static void extractFrameKeys(final Set<Key> retainedkeys, final Frame fr
-
-    /**
-     *
-     * @param retainedKeys Keys that are NOT deleted and will remain in DKV. Only Model keys and Frame keys are accepted
-     */
-    public ClearDKVTask(Key[] retainedKeys) {
-      _retainedKeys = retainedKeys;
+    private NodeKeysRemovalTask(final Key[] retainedKeys) {
+      _ignoredKeys = retainedKeys;
     }
 
     @Override
-    protected void setupLocal() {
-      final Set<Key> retainedKeys = new HashSet<>(_retainedKeys.length);
-      retainedKeys.addAll(Arrays.asList(_retainedKeys));
-
-      final Collection<Value> storeKeys = H2O.STORE.values();
-      Futures removalFutures = new Futures();
-      for (final Value value : storeKeys) {
-        if (retainedKeys.contains(value._key)) {
-          continue;
-        }
-        try {
-          // The `is*` methods of the Value class (isFrame, isModel() etc.) do call getFreezable method, which might
-          // throw ClassNotFoundException if the value has already been removed (type BAD). However, the methods do not
-          // declare the ClassNotFound exception, thefore the catch block does not compile. To avoid catching general
-          // Exception here,  the `TypeMap.getTheFreezableOrThrow` is called beforehand, as this method declares the 
-          // exception and the code compiles.
-          TypeMap.getTheFreezableOrThrow(value.type());
-
-          if (value.isNull()) continue;
-          if (value.isFrame()) {
-            ((Frame) value.get()).retain(removalFutures, retainedKeys);
-          } else if (value.isModel()) {
-            ((Model) value.get()).remove(removalFutures);
-          }
-        } catch (ClassNotFoundException e) {
-          // Keys are collected globally and then, on each node, this ClearDKVTask is invoked. Local H2O.STORE on each node is
-          // deleted, as it might contain local-only keys. This strategy is used to prevent collecting keys from all nodes and then
-          // invoking "remove" on one node. When each node iterates over local keys, one key might be removed multiple times.
-          // This might result in ClassNotFoundException, as getTheFreezable might from TypeMap might throw this error.
-          Log.debug(e);
-          continue; // Value not present, ignore
-        }
+    public void run() {
+      final Set<Key> values = H2O.localKeySet();
+      final Set<Key> ignoredSet = new HashSet<>();
+      final Futures futures = new Futures();
+
+      for (final Key ignoredKey : _ignoredKeys) {
+        ignoredSet.add(ignoredKey);
       }
-      removalFutures.blockForPending();
-    }
 
+      for (final Key key : values) {
+        if (ignoredSet.contains(key)) continue; // Do not perform DKV.get at all if the key is to be ignored
+        if (key.isChunkKey()) continue; // Chunks are deleted with vectors & are not Keyed
+
+        final Value value = DKV.get(key);
+        if (value == null || value.isVecGroup() || value.isNull()) continue;
```

Comment:
VecGroups are meanth to be deleted as a part of `Vec` only. It'll `H2O.fail()` on an attempt to `remove_impl` a VecGroup. Not sure whether that's the best thing to do, but let's respect it for now, unless there is a strong reason to change the behavior.